ql/indexes/equityindex.cpp

```diff
@@ -0,0 +1,89 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*
+ Copyright (C) 2023 Marcin Rybacki
+
+ This file is part of QuantLib, a free-software/open-source library
+ for financial quantitative analysts and developers - http://quantlib.org/
+
+ QuantLib is free software: you can redistribute it and/or modify it
+ under the terms of the QuantLib license.  You should have received a
+ copy of the license along with this program; if not, please email
+ <quantlib-dev@lists.sf.net>. The license is also available online at
+ <http://quantlib.org/license.shtml>.
+
+ This program is distributed in the hope that it will be useful, but WITHOUT
+ ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ FOR A PARTICULAR PURPOSE.  See the license for more details.
+*/
+
+#include <ql/indexes/equityindex.hpp>
+#include <ql/settings.hpp>
+#include <utility>
+
+namespace QuantLib {
+
+    EquityIndex::EquityIndex(std::string name,
+                             Currency currency,
+                             Calendar fixingCalendar,
+                             Handle<YieldTermStructure> interest,
+                             Handle<YieldTermStructure> dividend)
+    : name_(std::move(name)), currency_(std::move(currency)),
+      interest_(std::move(interest)), dividend_(std::move(dividend)),
+      settlementCalendar_(std::move(fixingCalendar)) {
+
+        registerWith(interest_);
+        registerWith(dividend_);
+        registerWith(Settings::instance().evaluationDate());
+        registerWith(IndexManager::instance().notifier(EquityIndex::name()));
+    }
+
+    Rate EquityIndex::fixing(const Date& fixingDate, bool forecastTodaysFixing) const {
+
+        QL_REQUIRE(isValidFixingDate(fixingDate), "Fixing date " << fixingDate << " is not valid");
+
+        Date today = Settings::instance().evaluationDate();
+
+        if (fixingDate > today || (fixingDate == today && forecastTodaysFixing))
+            return forecastFixing(fixingDate);
+
+        if (fixingDate <= today) {
+            // today's fixing is required
+            // even without enforcing it
+            return pastFixing(fixingDate);
+        }
+
+        return forecastFixing(fixingDate);
```

Comment:
Is this ever reached?  Otherwise the above can be an `if/else`

Reply:
I will adjust it.